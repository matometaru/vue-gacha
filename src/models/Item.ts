export default class Item {
  name: string;
  rare: number;
  description: string;

  constructor(name: string, rare: number, description: string) {
    this.name = name;
    this.rare = rare;
    this.description = description;
  }
}