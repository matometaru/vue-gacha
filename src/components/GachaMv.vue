<template>
  <div class="gacha-mv">
    <slick ref="slick" :options="slickOptions" @reInit="reInit">
      <div v-for="(item, index) in items" :key="index" class="gacha-mv-list">
        <p class="gacha-mv-list-name">{{ item.name }}</p>
        <p class="gacha-mv-list-rare">{{ item.rare }}</p>
        <p class="gacha-mv-list-description">{{ item.description }}</p>
      </div>
    </slick>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue';
import Gacha from '@/models/Gacha'
import Item from '@/models/Item'

// slick
import Slick from 'vue-slick';
// import 'slick-carousel/slick/slick.css';
// import 'slick-carousel/slick/slick-theme.css';

export default Vue.extend({
  name: "GachaMv",
  components: { Slick },
  props: {
    gacha: Object as PropType<Gacha>,
    items: Array as PropType<Item[]>
  },
  data() {
    return {
      slickOptions: {
        autoplay: true,
        dots: true,
        infinite: true,
        speed: 300,
        slidesToShow: 1,
        slidesToScroll: 1,
      }
    };
  },
  computed: {
    pickupItems(): Item[] {
      return this.items.filter((item) => {
        return item.isPickup();
      });
    }
  },
  watch: {
    items() {
      (this.$refs.slick as any).destroy();
      this.$nextTick(() => {
        (this.$refs.slick as any).create();
      });
    }
  },
  methods: {
    reInit() {
      this.$nextTick(() => {
          (this.$refs.slick as any).reSlick();
      });
    },
  }
});
</script>

<style scoped lang="scss">
.gacha-center {
  margin: 40px 0 0;
  &-header {
    font-size: 20px;
  }
}
</style>
