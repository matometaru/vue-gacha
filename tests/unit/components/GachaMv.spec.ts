import { shallowMount } from "@vue/test-utils";
import GachaMv from "@/components/GachaMv.vue";

import Item from '@/models/Item'
import Slick from 'vue-slick';

describe("GachaMv.vue", () => {
  it("itemsの値がhtmlに出力されているか？", () => {
    const items = [
      new Item("アイテム名", 3, "アイテム説明")
    ];
    const wrapper = shallowMount(GachaMv, {
      propsData: { items }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("slickの初期値オプション通りのhtmlに出力されているか？", async() => {
    const items = [
      new Item("アイテム名3", 3, "アイテム説明3"),
      new Item("アイテム名4", 4, "アイテム説明4"),
    ];
    const wrapper = shallowMount(GachaMv, {
      stubs: {
        'slick': Slick,
      },
      propsData: { items }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  it("pickupItemが期待通りか？", async() => {
    const items = [
      new Item("アイテム名3", 3, "アイテム説明3"),
      new Item("アイテム名4", 4, "アイテム説明4"),
      new Item("アイテム名5", 5, "アイテム説明5"),
    ];
    const wrapper = shallowMount(GachaMv, {
      propsData: { items }
    });
    expect((wrapper.vm as any).pickupItems).toEqual([new Item("アイテム名5", 5, "アイテム説明5")]);
  });
});
