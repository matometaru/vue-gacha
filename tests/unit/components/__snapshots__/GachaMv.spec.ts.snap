// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GachaMv.vue itemsの値がhtmlに出力されているか？ 1`] = `
<div class="gacha-mv">
  <slick-stub options="[object Object]">
    <div class="gacha-mv-list">
      <p class="gacha-mv-list-name">アイテム名</p>
      <p class="gacha-mv-list-rare">3</p>
      <p class="gacha-mv-list-description">アイテム説明</p>
    </div>
  </slick-stub>
</div>
`;

exports[`GachaMv.vue slickの初期値オプション通りのhtmlに出力されているか？ 1`] = `
<div class="gacha-mv">
  <div class="slick-initialized slick-slider slick-dotted"><button class="slick-prev slick-arrow" aria-label="Previous" type="button" style="">Previous</button>
    <div class="slick-list draggable">
      <div class="slick-track" style="opacity: 1; width: 0px; transform: translate(0px, 0px);">
        <div class="slick-slide slick-cloned" data-slick-index="-1" id="" aria-hidden="true" style="width: 0px;" tabindex="-1">
          <div>
            <div class="gacha-mv-list" style="width: 100%; display: inline-block;">
              <p class="gacha-mv-list-name">アイテム名4</p>
              <p class="gacha-mv-list-rare">4</p>
              <p class="gacha-mv-list-description">アイテム説明4</p>
            </div>
          </div>
        </div>
        <div class="slick-slide slick-current slick-active" data-slick-index="0" aria-hidden="false" style="margin-left: 0px; width: 0px;" role="tabpanel" id="slick-slide00">
          <div>
            <div class="gacha-mv-list" style="width: 100%; display: inline-block;">
              <p class="gacha-mv-list-name">アイテム名3</p>
              <p class="gacha-mv-list-rare">3</p>
              <p class="gacha-mv-list-description">アイテム説明3</p>
            </div>
          </div>
        </div>
        <div class="slick-slide" data-slick-index="1" aria-hidden="true" style="width: 0px;" tabindex="-1" role="tabpanel" id="slick-slide01">
          <div>
            <div class="gacha-mv-list" style="width: 100%; display: inline-block;">
              <p class="gacha-mv-list-name">アイテム名4</p>
              <p class="gacha-mv-list-rare">4</p>
              <p class="gacha-mv-list-description">アイテム説明4</p>
            </div>
          </div>
        </div>
        <div class="slick-slide slick-cloned" data-slick-index="2" id="" aria-hidden="true" style="width: 0px;" tabindex="-1">
          <div>
            <div class="gacha-mv-list" style="width: 100%; display: inline-block;">
              <p class="gacha-mv-list-name">アイテム名3</p>
              <p class="gacha-mv-list-rare">3</p>
              <p class="gacha-mv-list-description">アイテム説明3</p>
            </div>
          </div>
        </div>
        <div class="slick-slide slick-cloned" data-slick-index="3" id="" aria-hidden="true" style="width: 0px;" tabindex="-1">
          <div>
            <div class="gacha-mv-list" style="width: 100%; display: inline-block;">
              <p class="gacha-mv-list-name">アイテム名4</p>
              <p class="gacha-mv-list-rare">4</p>
              <p class="gacha-mv-list-description">アイテム説明4</p>
            </div>
          </div>
        </div>
      </div>
    </div><button class="slick-next slick-arrow" aria-label="Next" type="button" style="">Next</button>
    <ul class="slick-dots" style="" role="tablist">
      <li class="slick-active" role="presentation"><button type="button" role="tab" id="slick-slide-control00" aria-controls="slick-slide00" aria-label="1 of 2" tabindex="0" aria-selected="true">1</button></li>
      <li role="presentation"><button type="button" role="tab" id="slick-slide-control01" aria-controls="slick-slide01" aria-label="2 of 2" tabindex="-1">2</button></li>
    </ul>
  </div>
</div>
`;
